File: src/utils/errorHandler.js
/**
 * Centralized Error Handling
 * Custom error classes + Express error-handling middleware.
 */

// ------- Custom Error Classes -------

class AppError extends Error {
  constructor(message, statusCode = 500, details = null) {
    super(message);
    this.name = this.constructor.name;
    this.statusCode = statusCode;
    this.details = details;
    this.isOperational = true; // distinguishes expected errors from bugs
    Error.captureStackTrace(this, this.constructor);
  }
}

class ValidationError extends AppError {
  constructor(message, details = null) {
    super(message, 400, details);
  }
}

class NotFoundError extends AppError {
  constructor(resource = 'Resource') {
    super(`${resource} not found`, 404);
  }
}

class ExternalServiceError extends AppError {
  constructor(service, message) {
    super(`${service} error: ${message}`, 502);
  }
}

class RateLimitError extends AppError {
  constructor(retryAfter = 60) {
    super('Rate limit exceeded, please try again later', 429);
    this.retryAfter = retryAfter;
  }
}

// ------- Express Error-Handling Middleware -------

/**
 * Global error handler — must be registered LAST in the middleware chain.
 * Signature: (err, req, res, next)
 */
function errorHandler(err, req, res, _next) {
  // Log every error with timestamp
  const timestamp = new Date().toISOString();
  console.error(`[${timestamp}] ERROR:`, {
    name: err.name,
    message: err.message,
    statusCode: err.statusCode || 500,
    path: req.originalUrl,
    method: req.method,
    ...(process.env.NODE_ENV !== 'production' && { stack: err.stack })
  });

  // Operational errors → send structured response
  if (err.isOperational) {
    return res.status(err.statusCode).json({
      success: false,
      error: {
        message: err.message,
        ...(err.details && { details: err.details }),
        ...(err.retryAfter && { retryAfter: err.retryAfter })
      }
    });
  }

  // Multer file-size error
  if (err.code === 'LIMIT_FILE_SIZE') {
    return res.status(413).json({
      success: false,
      error: { message: 'File size exceeds the 50 MB limit' }
    });
  }

  // Joi validation error (if thrown directly)
  if (err.isJoi || err.name === 'ValidationError') {
    return res.status(400).json({
      success: false,
      error: {
        message: 'Validation failed',
        details: err.details || err.message
      }
    });
  }

  // Unknown / programming errors → generic 500
  res.status(500).json({
    success: false,
    error: {
      message:
        process.env.NODE_ENV === 'production'
          ? 'Internal server error'
          : err.message
    }
  });
}

/**
 * Wrap an async route handler so thrown errors reach the error middleware.
 * Usage: router.post('/upload', asyncHandler(controller.upload));
 */
function asyncHandler(fn) {
  return (req, res, next) => {
    Promise.resolve(fn(req, res, next)).catch(next);
  };
}

module.exports = {
  AppError,
  ValidationError,
  NotFoundError,
  ExternalServiceError,
  RateLimitError,
  errorHandler,
  asyncHandler
};

File: app/utils/error_handler.py
"""
Centralized Error Handling
Custom exception classes + FastAPI exception handlers.
"""
from fastapi import Request
from fastapi.responses import JSONResponse


# ─── Custom Exception Classes ──────────────────────────────

class AppError(Exception):
    """Base application error."""
    def __init__(self, message: str, status_code: int = 500, details=None):
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.details = details


class ValidationError(AppError):
    """Input validation failed."""
    def __init__(self, message: str, details=None):
        super().__init__(message, 400, details)


class NotFoundError(AppError):
    """Requested resource not found."""
    def __init__(self, resource: str = "Resource"):
        super().__init__(f"{resource} not found", 404)


class ExternalServiceError(AppError):
    """Third-party service (Grok, ChromaDB) failure."""
    def __init__(self, service: str, message: str):
        super().__init__(f"{service} error: {message}", 502)


class RateLimitError(AppError):
    """API rate limit exceeded."""
    def __init__(self, retry_after: int = 60):
        super().__init__("Rate limit exceeded, please try again later", 429)
        self.retry_after = retry_after


# ─── FastAPI Exception Handlers ────────────────────────────

async def app_error_handler(request: Request, exc: AppError) -> JSONResponse:
    """Handle all custom AppError subclasses."""
    body: dict = {
        "success": False,
        "error": {"message": exc.message},
    }
    if exc.details:
        body["error"]["details"] = exc.details
    if isinstance(exc, RateLimitError):
        body["error"]["retryAfter"] = exc.retry_after
    return JSONResponse(status_code=exc.status_code, content=body)


async def generic_error_handler(request: Request, exc: Exception) -> JSONResponse:
    """Catch-all for unhandled exceptions."""
    return JSONResponse(
        status_code=500,
        content={
            "success": False,
            "error": {"message": "Internal server error"},
        },
    )
